### ISSUE TITLE
OPERA-MS

### ISSUE THREAD

**Issue by liluacrobat** [ubccr/software-layer#306]
The EB recipe for Opera-MS is here:  [https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/o/OPERA-MS/OPERA-MS-0.9.0-20200802-foss-2020a-Python-3.8.2.eb](https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Feasybuilders%2Feasybuild-easyconfigs%2Fblob%2Fdevelop%2Feasybuild%2Feasyconfigs%2Fo%2FOPERA-MS%2FOPERA-MS-0.9.0-20200802-foss-2020a-Python-3.8.2.eb&data=05%7C02%7Clli59%40g-mail.buffalo.edu%7C4849bbacf24f447b36c408dc6dd5fdfc%7C96464a8af8ed40b199e25f6b50a20250%7C0%7C0%7C638506013062183339%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=6K7xlF0FrqFCWsCy0X4LU9hGwtgwV4hElNxL8a%2FhzJw%3D&reserved=0)

**Comment by tonykew** [comment]
I can build its dependent module Statistics-R version 0.34, but OPERA-MS-0.9.0-20200802
fails the third test of nine, a coverage_estimation test.
The rather odd versioning number of the module is because "OPERA-MS-0.9.0" is a git branch
which is (surprisingly) six years ahead of the "master" branch.   The "20200802" is a date for
a specific git commit 'c18b4f3' from that date.
I created a new .eb file using the most recent git commit from the "OPERA-MS-0.9.0" git
branch, that is git commit '363957c' from December 12th 2023 - this fails the same test in
the same way.

I have no way to fix this.

The only thing I could do it try skipping the tests, but the resulting module may not work as
expected, or at all.

**Comment by liluacrobat** [comment]
Could we skip the tests? I complied the original code with gcc/9.3.0. But to run the program, I need to load gcc/11.2.0. Following is a case that could work for the coverage_estimate by loading existing models. But it will fail in the last step of gap_filling. I'm not sure whether the code has something wrong or it is becasue the code is complied with a different gcc version. If skip gap_filling, the program could successfully complete. Hope it could help.

module load ccrsoft/2023.01
module load java/11.0.16
module load gcc/9.3.0
module load perl
module load gcc/11.2.0  openmpi/4.1.1
module load scipy-bundle/2021.10
module load pandas/1.3.4

export PATH=$PATH:/cvmfs/soft.ccr.buffalo.edu/versions/2023.01/easybuild/software/avx512/MPI/gcc/11.2.0/openmpi/4.1.1/r/4.2.0/bin
export PERL5LIB="/user/lli59/perl5/lib/perl5${PERL5LIB:+:${PERL5LIB}}"

echo '--------------------'
echo 'Begin ...'

OPERA_MS=/projects/academic/pidiazmo/projectsoftwares/OPERA-MS_R2/OPERA-MS_s2
perl $OPERA_MS/OPERA-MS.pl \
--short-read1 /projects/academic/pidiazmo/lu/Mock_Community/opera_test/fastq_short/BEN_Half_R1_001.fastq \
--short-read2 /projects/academic/pidiazmo/lu/Mock_Community/opera_test/fastq_short/BEN_Half_R2_001.fastq \
--long-read /projects/academic/pidiazmo/lu/Mock_Community/opera_test/fastq_long/BEN_Half.fastq \
    --genome-db /projects/academic/pidiazmo/projectsoftwares/OPERA-MS/HOMD/v10.1/opera_homd_r10_1_v3 \
--out-dir BEN_Half_RESULTS \
    --num-processors 12 \
    --no-polishing \
    --no-gap-filling

**Comment by tonykew** [comment]
I'll try to skip the tests.
Can you provide a test I can use to make sure the build is functional, as least for you?
Thanks

**Comment by liluacrobat** [comment]
It failed in coverage_estimation is because `GLIBCXX_3.4.29' is not included in libstdc++.so.6 in latter version of gcc. If complied with, for example, gcc/11.2.0 it will fail. If the code is compiled with gcc/9.3.0. It will succeed.

Following is a successful case of mine:
module load ccrsoft/2023.01
module load java/11.0.16
module load gcc/9.3.0
module load perl
module load gcc/11.2.0 openmpi/4.1.1
module load scipy-bundle/2021.10
module load pandas/1.3.4

export PATH=$PATH:/cvmfs/soft.ccr.buffalo.edu/versions/2023.01/easybuild/software/avx512/MPI/gcc/11.2.0/openmpi/4.1.1/r/4.2.0/bin
export PERL5LIB="/user/lli59/perl5/lib/perl5${PERL5LIB:+:${PERL5LIB}}"

OPERA_MS=/projects/academic/pidiazmo/projectsoftwares/OPERA-MS_R2/OPERA-MS_s2

perl $OPERA_MS/OPERA-MS.pl\
    --contig-file $OPERA_MS/test_files/contigs.fasta\
    --short-read1 $OPERA_MS/test_files/R1.fastq.gz\
    --short-read2 $OPERA_MS/test_files/R2.fastq.gz\
    --long-read $OPERA_MS/test_files/long_read.fastq\
    --no-ref-clustering \
    --no-polishing \
    --no-gap-filling \
    --out-dir test_files/RESULTS 2> test_files/log.err

But I cannot make it work with gap-filing or polishing. It will fail with gap-filling
perl $OPERA_MS/OPERA-MS.pl\
    --contig-file $OPERA_MS/test_files/contigs.fasta\
    --short-read1 $OPERA_MS/test_files/R1.fastq.gz\
    --short-read2 $OPERA_MS/test_files/R2.fastq.gz\
    --long-read $OPERA_MS/test_files/long_read.fastq\
    --no-ref-clustering \
    --out-dir test_files/RESULTS 2> test_files/log.err

**Comment by tonykew** [comment]
My build problems do not relate to the GCC version - certainly using gcc/9.3.0 when the
rest of the ccrsoft/2023.01 software release was compiled with gcc 11.2.0 wlll lead to
inevitable problems.
I have identified a library problem with my build, and I'm working on that now.
Thanks for providing a test that should work.

More into to follow...

**Comment by tonykew** [comment]
I fixed the library issue, fixed MUMmer3.23 that the build installs and fixed the PERL5LIB path
for MUMmer3.23
Now the dependency verification completes:

```
$ OPERA-MS check-dependency
 *** OPERA-LG functional
 *** sigma functional
 *** samtools functional
 *** blasr functional
 *** bwa functional
 *** minimap2 functional
 *** mash functional
 *** mummer functional
 *** megahit functional
 *** spades functional
 *** racon functional
 *** pilon functional

 *** All compiled software are functional.
 *** Please try to run OPERA-MS on the test dataset.
$ 
```

...and I can run the test with gap-filing and polishing:

```
$ rm -rf OUTPUT/
$ OPERA-MS -contig-file test_files/contigs.fasta \
 --short-read1 test_files/R1.fastq.gz \
 --short-read2 test_files/R2.fastq.gz \
 --long-read test_files/long_read.fastq \
 --no-ref-clustering \
 --out-dir ./OUTPUT
[...]
[Thu May 16 23:33:51 2024]	OPERA-MS completed
$ 
```


I will submit this version for internal testing and update this request
once the software is published to the ccrsoft/2023.01 software release/
